girder/api/v1/resource.py

```diff
@@ -139,6 +140,124 @@ def _getResourceModel(self, kind, funcName=None):
+                break
+
+        if candidateParent is None:
+            raise RestException(
+                'Parent resource not found: {}'.format(parentId))
+
+        candidateChild = self.model('folder').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'folder'
+
+        filterObject['folderId'] = filterObject.pop('parentId')
+        del filterObject['baseParentId']
+        del filterObject['baseParentType']
+        candidateChild = self.model('item').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'item'
+
+        filterObject['itemId'] = filterObject.pop('folderId')
+        candidateChild = self.model('file').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'file'
+
+        raise RestException('Child resource not found: {}->{}'.format(
+            parentId, token))
+
+    def _lookupPath(self, path, user=None):
+        """
+        Find a particular resource by path or throw an exception.  The given
+        user must have read access to all resources featured in the path.
+        :param path: the path of the object to find
+        :param user: the user to search as
+        :returns: the resource.
+        """
+
+        if not path.startswith('/'):
+            raise RestException('Invalid path format')
+
+        pathArray = path[1:].split('/')
+        model = pathArray[0]
+        isUserPath = (model == 'user')
+        isCollectionPath = (model == 'collection')
+
+        if user is None:
+            user = self.getCurrentUser()
+
+        parent = None
+        if isUserPath:
+            username = pathArray[1]
+            parent = self.model('user').findOne({'login': username})
+
+            if parent is None:
+                raise RestException('User not found: {}'.format(username))
+
+        elif isCollectionPath:
+            collectionName = pathArray[1]
+            parent = self.model('collection').findOne({'name': collectionName})
+
+            if parent is None:
+                raise RestException(
+                    'Collection not found: {}'.format(collectionName))
+
+        else:
+            raise RestException('Invalid path format')
+
+        try:
+            document = parent
+            self.model(model).requireAccess(document, user)
+            for token in pathArray[2:]:
+                document, model = self._lookupToken(token, document['_id'])
+                self.model(model).requireAccess(document, user)
+
+        except RestException:
+            raise RestException('Path not found: {}'.format(path))
+
+        result = self.model(model).filter(document, user)
+        return result
+
+    @access.public
+    def stat(self, params):
+        path = params.get('path')
+        currentUser = self.getCurrentUser()
+        result = self._lookupPath(path, currentUser)
+        return result
```

Comment:
Since LOC ∝ n_bugs, I vote we rewrite this method as

```
self.requireParams('path', params)
return self._lookupPath(params['path'], self.getCurrentUser())
```

(I included the requireParams call since currently a confusing exception would bubble up in cases where someone forgot to pass the path parameter.)